### PR TITLE
WIP - Alerts assignment

### PR DESCRIPTION
**Pull request by Omri-Levy**
- feat(backoffice-v2): added ability to select rows
- feat(backoffice-v2): now disabling assign button if no alerts are selected
- feat(backoffice-v2): added a generic dropdown component
- feat(*): added the ability to bulk assign and unassign

image
image
image

### CODE REVIEW

File: apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { DropdownMenu } from '@/common/components/molecules/DropdownMenu/DropdownMenu';
+import { DropdownMenuTrigger } from '@/common/components/molecules/DropdownMenu/DropdownMenu.Trigger';
+import { DropdownMenuContent } from '@/common/components/molecules/DropdownMenu/DropdownMenu.Content';
+import { isNonEmptyArray } from '@ballerine/common';
+import { DropdownMenuItem } from '@/common/components/molecules/DropdownMenu/DropdownMenu.Item';
+import React from 'react';
+import { DropdownProps } from '@/common/components/molecules/Dropdown/interfaces';
+
+export const Dropdown = <TItem,>({ options, trigger, props, children }: DropdownProps<TItem>) => {
```
Comment (liorzam) [inline apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx:9]: we dont have any existing dropdown component already?

**Review by alonp99 — APPROVED** [review]

File: apps/backoffice-v2/src/domains/alerts/fetchers.ts [hunk 2]
```diff
@@ -79,3 +83,23 @@ export const fetchAlerts = async (params: {
...
+  alertIds,
+}: {
+  assigneeId: string | null;
+  alertIds: string[];
+}) => {
+  const [alerts, error] = await apiClient({
+    url: `${getOriginUrl(env.VITE_API_URL)}/api/v1/external/alerts/assign`,
+    method: Method.PATCH,
+    body: {
+      assigneeId,
```
Comment (liorzam) [inline apps/backoffice-v2/src/domains/alerts/fetchers.ts:98]: you should also pass projectid

File: apps/backoffice-v2/src/domains/alerts/hooks/mutations/useAssignAlertsMutation/useAssignAlertsMutation.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  const { data: users } = useUsersQuery();
+
+  return useMutation({
+    mutationFn: ({
+      assigneeId,
+      alertIds,
+    }: {
+      assigneeId: string | null;
+      alertIds: string[];
+      isAssignedToMe: boolean;
```
Comment (liorzam) [inline apps/backoffice-v2/src/domains/alerts/hooks/mutations/useAssignAlertsMutation/useAssignAlertsMutation.tsx:45]: why does it matter?
in the component which is assign to the same assigneeId you just set another label ... 
i think its redundant to keep such a property

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx [hunk 4]
```diff
@@ -21,21 +25,49 @@ export const useAlertsTableLogic = ({ data }: { data: TAlertsList }) => {
   ]);
   const onSortingChange: OnChangeFn<SortingState> = useCallback(
     updaterOrValue => {
-      setSorting(updaterOrValue);
+      setSorting(old => {
+        if (!isInstanceOfFunction(updaterOrValue)) {
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: can you explain what is isInstanceOfFunction checks?

File: apps/backoffice-v2/src/common/components/atoms/IndeterminateCheckbox/IndeterminateCheckbox.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  checked,
+  indeterminate,
+  className,
+  onCheckedChange,
+  ...rest
+}) => {
+  return (
+    <Checkbox_
+      className={ctw(className)}
+      checked={!checked && indeterminate ? 'indeterminate' : checked}
```
Comment (chesterkmr) [inline apps/backoffice-v2/src/common/components/atoms/IndeterminateCheckbox/IndeterminateCheckbox.tsx:19]: Consider separate `checked` and `indeterminate`

Good [reference](https://mui.com/material-ui/react-checkbox/#indeterminate)

File: apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  return (
+    <DropdownMenu {...props?.menu}>
+      <DropdownMenuTrigger {...props?.trigger}>{trigger}</DropdownMenuTrigger>
+      <DropdownMenuContent {...props?.content}>
+        <>
+          {isNonEmptyArray(options) &&
+            options.map(item => children?.({ item, DropdownItem: DropdownMenuItem }))}
+          {!isNonEmptyArray(options) && <DropdownMenuItem>No options</DropdownMenuItem>}
```
Comment (chesterkmr) [inline apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx:17]: Same check performed two times, ternary operator won't fit here?

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx [hunk 7]
```diff
@@ -21,21 +25,49 @@ export const useAlertsTableLogic = ({ data }: { data: TAlertsList }) => {
   ]);
   const onSortingChange: OnChangeFn<SortingState> = useCallback(
     updaterOrValue => {
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: can you change the name to `sortingUpdater`?

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx [hunk 8]
```diff
@@ -21,21 +25,49 @@ export const useAlertsTableLogic = ({ data }: { data: TAlertsList }) => {
...
-      setSorting(updaterOrValue);
+      setSorting(old => {
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: change old -> `oldSortingVal`

**Review by chesterkmr — APPROVED** [review]
LGTM!

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx [hunk 9]
```diff
@@ -21,21 +25,49 @@ export const useAlertsTableLogic = ({ data }: { data: TAlertsList }) => {
...
+            sortBy: updaterOrValue[0]?.id || 'dataTimestamp',
+            sortDir: updaterOrValue[0]?.desc ? 'desc' : 'asc',
+          });
 
-      if (!isInstanceOfFunction(updaterOrValue)) return;
+          return updaterOrValue;
+        }
 
-      const [currentSorting] = updaterOrValue();
+        const nextValue = updaterOrValue(old);
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: nextValue can also change to `newSortVal`

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            item?.value?.id === authenticatedUserId,
+          )}
+        >
+          {item?.value === null && (
+            <>
+              <UnassignedAvatarSvg className="d-[1.375rem]" />
+              <span>Unassign</span>
+            </>
+          )}
+          {item?.value !== null && (
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx:74]: how could be a case an item without a value property?

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  isAssignedToMe: boolean;
+  onAssigneeSelect: (id: string | null, isAssignedToMe: boolean) => () => void;
+}> = ({ assignees, authenticatedUserId, isDisabled, onAssigneeSelect }) => {
+  const sortedAssignees = useMemo(
+    () =>
+      // Sort assignees so that the authenticated user is always first
+      assignees
+        ?.slice()
+        ?.sort((a, b) =>
+          a?.id === authenticatedUserId ? -1 : b?.id === authenticatedUserId ? 1 : 0,
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx:20]: why dont we sort the assignees by A-Z?

File: apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx (hunk 1 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx:9]: It was requested that we'll start creating components that abstract ShadCN's multiple pieces. The dropdown we used until now was `DropdownMenu.Item`, `DropdownMenu.Content` and so on inline.

File: apps/backoffice-v2/src/domains/alerts/fetchers.ts (hunk 2 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/domains/alerts/fetchers.ts:98]: @alonp99 wants project ids to be handled by the server.

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx (hunk 4 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: `updaterOrValue` can be a function. At a minimum this is to appease TypeScript.

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx (hunk 7 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsTable/hooks/useAlertsTableLogic/useAlertsTableLogic.tsx]: You mean something like this?
```suggestion
    sortingUpdaterOrValue => {
```

File: apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx (hunk 6 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/common/components/molecules/Dropdown/Dropdown.tsx:17]: I find ternaries in JSX ruin formatting and makes it hard to read and follow. I am considering `ts-pattern` though. 🤤

File: apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      {
+        id: 'unassign',
+        value: null,
+      },
+      ...(sortedAssignees?.map(assignee => ({
+        id: assignee?.id,
+        value: assignee,
+      })) ?? []),
+    ];
+  }, [sortedAssignees]);
```
Comment (liorzam) [inline apps/backoffice-v2/src/pages/TransactionMonitoringAlerts/components/AlertsAssignDropdown/AlertsAssignDropdown.tsx:36]: Im a bit confusing why our component cant work well with the original assignees list?

im not familiar with a convention of having object in a value property.

Usually when an event is occurred by the component like onClick/OnChange then the item value will be present on `e.target.value`

**Comment by liorzam** [comment]
Add new tech debt task to use our new generic dropdown component

File: apps/backoffice-v2/src/domains/alerts/hooks/mutations/useAssignAlertsMutation/useAssignAlertsMutation.tsx (hunk 3 above)
Comment (liorzam) [reply apps/backoffice-v2/src/domains/alerts/hooks/mutations/useAssignAlertsMutation/useAssignAlertsMutation.tsx:45]: Lets talk about it in tech debt please add a task and specify this comment link in the description

File: apps/backoffice-v2/src/domains/alerts/fetchers.ts (hunk 2 above)
Comment (liorzam) [reply apps/backoffice-v2/src/domains/alerts/fetchers.ts:98]: @alonp99  @nitzanballerine How can we determine what is the selected project id from application perspective ?

Comment (alonp99) [reply apps/backoffice-v2/src/domains/alerts/fetchers.ts:98]: In the future we can use what you did with the query param - but we don't need it right away - to reduce scope we can choose the authenticated user project id by default || query params project ids

**Review by liorzam — APPROVED** [review]
Alt Text
Awesome!

**Pull request merged**